src/Interpreter.purs

```diff
@@ -100,38 +98,97 @@ runAsm2 = interpret >>> exec
     Tuple (Left err) s → Console.log $ show err <> "\n " <> show s
     Tuple (Right _unit) s → Console.logShow s
 
-newtype AsmM3 a = AsmM3 (Effect (Ref (Maybe Error /\ St /\ a)))
+newtype AsmM3 a = AsmM3 (Effect (Ref (ExceptT Error (Tuple St) a)))
 
 derive instance Newtype (AsmM3 a) _
 
 instance Functor AsmM3 where
   map f asm = wrap do
-    ra ← unwrap asm
-    Tuple error (Tuple st a) ← Ref.read ra
-    Ref.new (Tuple error (Tuple st (f a)))
+    ea ← Ref.read =<< unwrap asm
+    Ref.new $ f <$> ea
 
 instance Apply AsmM3 where
-  apply = ?todo
+  apply ∷ ∀ a b. AsmM3 (a → b) → AsmM3 a → AsmM3 b
+  apply = ap
 
 instance Applicative AsmM3 where
-  pure = wrap <<< Ref.new <<< Tuple Nothing <<< Tuple Map.empty
+  pure ∷ ∀ a. a → AsmM3 a
+  pure = wrap <<< Ref.new <<< wrap <<< Tuple Map.empty <<< pure
 
 instance Bind AsmM3 where
   bind ∷ ∀ a b. AsmM3 a → (a → AsmM3 b) → AsmM3 b
-  bind = ?homework
+  bind asm f = wrap do
+    unwrap asm >>= Ref.read >>= runExceptT >>> case _ of
+      Tuple sta (Left err) → Ref.new $ wrap $ Tuple sta $ Left err
+      Tuple sta (Right a) → do
+        unwrap (f a) >>= Ref.read >>= runExceptT >>> case _ of
+          Tuple stb (Left err) →
+            let
+              st' = Map.unionWith (flip const) sta stb
+            in
+              Ref.new $ wrap $ Tuple st' $ Left err
+          Tuple stb (Right b) →
+            let
+              st' = Map.unionWith (flip const) sta stb
+            in
+              Ref.new $ wrap $ Tuple st' $ Right b
 
 instance Monad AsmM3
 
-instance MonadState St AsmM3 where
-  state ∷ ∀ a. (St → Tuple a St) → AsmM3
-  state f = ?homeworkState
-
 instance MonadThrow Error AsmM3 where
   throwError ∷ ∀ a. Error → AsmM3 a
-  throwError = ?throwError
+  throwError = wrap <<< Ref.new <<< wrap <<< Tuple Map.empty <<< Left
 
 runAsm3 ∷ Asm → Effect Unit
-runAsm3 = interpret >>> exec
+runAsm3 asm = do
+  r ← Ref.new $ Tuple (Map.empty ∷ Map Reg Int) (Nothing ∷ Maybe Error)
+  go r asm
   where
-  exec ∷ AsmM3 → Effect Unit
-  exec = ?exec
+  go r = case _ of
+    Set i reg next → do
+      Tuple st mbError ← Ref.read r
+      case mbError of
+        Just err →
+          showErr err st
+        Nothing → do
```

Comment:
I can see that this code is repeated 4 times. Please try to extract another function and replace repeated code by  its usage ("make code [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)")